### PR TITLE
Pytorch Device handling

### PR DESCRIPTION
**Pull request by le1nux**
* DatasetBatch and InferenceResultBatch have now device methods such as detach, to_device, get_device, to_cpu
* Reduced  GPU memory footprint in Trainer and Evaluator by moving respective batches to CPU directly after instantiation.

### CODE REVIEW

File: pytest/batching/test_batch.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    @pytest.fixture
+    def inference_batch_result(self) -> InferenceResultBatch:
+        targets = torch.IntTensor([0, 0, 0, 1, 1, 1])
+        tags = torch.IntTensor([0, 0, 0, 1, 1, 1])
+        predictions = torch.FloatTensor([0, 0.1, 0.01, 0.9, 0.7, 1])
+        return InferenceResultBatch(targets={TestInferenceResultBatch.target_key: targets},
+                                    predictions={TestInferenceResultBatch.prediction_key: predictions},
+                                    tags=tags)
+
+    # test device operations
```
Comment (marenpielka) [inline pytest/batching/test_batch.py:19]: Without individual assertions, those tests are a bit redundant or?

File: src/ml_gym/gym/trainer.py [hunk 2]
```diff
@@ -48,19 +48,22 @@ def check_if_initiable_component(loss_fun: Loss):
             self.logger.log(LogLevel.INFO, "Running warmup...")
             with torch.no_grad():
                 prediction_batches = self.map_batches(fun=self.forward_batch,
-                                                      fun_params={"device": device, "model": model},
+                                                      fun_params={"calculation_device": device, "model": model,
+                                                                  "result_device": torch.device("cpu")},
```
Comment (marenpielka) [inline src/ml_gym/gym/trainer.py:52]: Should the result_device be hardcoded? We could also take it from the config or?

File: pytest/batching/test_batch.py (hunk 1 above)
Comment (le1nux) [reply pytest/batching/test_batch.py:19]: yeah they run "at least" through then. In the deployment environment we unfortunately don't have any GPUs to test with.

File: src/ml_gym/gym/trainer.py (hunk 2 above)
Comment (le1nux) [reply src/ml_gym/gym/trainer.py:52]: When we create the InferenceResultBatch, the tensors don't have to be on the GPU anymore, because the training step is done. It's just used to calculates metrics etc. from that point on.
